File: combinatorics/resource/ST/project_5/bai_toan_11_12_13/dfs_general.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

void input_graph(int& n, vector<vector<int>>& adj, bool& directed) {
    int m;
    cout << "Nhap so dinh: ";
    cin >> n;
    cout << "Nhap so canh: ";
    cin >> m;
    cout << "Do thi co huong? (y/n): ";
    string s; cin >> s;
    directed = (s == "y" || s == "Y");
    adj.resize(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cout << "Nhap canh u v: ";
        cin >> u >> v;
        adj[u].push_back(v);
        if (!directed)
            adj[v].push_back(u);
    }
}

void dfs(int u, const vector<vector<int>>& adj, vector<bool>& visited) {
    visited[u] = true;
    cout << u << " ";
    for (int v : adj[u]) {
        if (!visited[v])
            dfs(v, adj, visited);
    }
}

int main() {
    int n;
    bool directed;
    vector<vector<int>> adj;
    input_graph(n, adj, directed);
    vector<bool> visited(n, false);
    cout << "Cac thanh phan lien thong:\n";
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            dfs(i, adj, visited);
            cout << endl;
        }
    }
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_11_12_13/dfs_general.py
def input_graph():
    n = int(input("Nhap so dinh: "))
    m = int(input("Nhap so canh: "))
    directed = input("Do thi co huong? (y/n): ").strip().lower() == 'y'
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u, v = map(int, input("Nhap canh u v: ").split())
        adj[u].append(v)
        if not directed:
            adj[v].append(u)
    return n, adj

def dfs(u, adj, visited):
    visited[u] = True
    print(u, end=" ")
    for v in adj[u]:
        if not visited[v]:
            dfs(v, adj, visited)

if __name__ == "__main__":
    n, adj = input_graph()
    visited = [False] * n
    print("Cac thanh phan lien thong:")
    for i in range(n):
        if not visited[i]:
            dfs(i, adj, visited)
            print()